sygnal/webpushpushkin.py

```diff
@@ -186,15 +199,17 @@ async def _dispatch_notification_unlimited(self, n, device, context):
         try:
             with SEND_TIME_HISTOGRAM.time():
                 with ACTIVE_REQUESTS_GAUGE.track_inprogress():
-                    response_wrapper = webpush(
+                    request = webpush(
                         subscription_info=subscription_info,
                         data=data,
                         ttl=self.ttl,
                         vapid_private_key=self.vapid_private_key,
                         vapid_claims=vapid_claims,
-                        requests_session=self.http_agent_wrapper,
+                        requests_session=self.http_request_factory,
+                    )
+                    response = await request.execute(
+                        self.http_agent, low_priority, topic
                     )
```

Comment:
where does the deduplication actually take place, out of interest?

Reply:
We don't do this, it happens in the push gateway. If you have two messages with the same topic header, the second will replace the first (if the first hasn't been delivered yet).

Reply:
ah, so there is a separate "push gateway" after sygnal (which I tend to think of as a, y'know, push gateway)?

Reply:
I mean, webpush pipes notification to a push provider (or gateway, hard to name things correctly with so many parts to it) like FCM that will actually deliver it to the browser application, similarly to how the apns and gcm pushkins forward it to the push providers of those platforms. Every browser that supports web push needs to have a server like this that application servers can post push messages to. It's the browser-specific push provider that will drop the first message if it hasn't been deliver yet.